src/clp_ffi_py/ir/native/PyKeyValuePairLogEvent.cpp

```diff
@@ -65,6 +78,149 @@ class IrUnitHandler {
+ * Helper class for `PyKeyValuePairLogEvent::to_dict`.
+ */
+class PyDictSerializationIterator {
+public:
+    // Factory function
+    /**
+     * Creates an iterator with the given inputs.
+     * @param schema_tree_node
+     * @param schema_subtree_bitmap
+     * @param parent
+     * @return A newly created iterator that holds a new reference of a Python dictionary on
+     * success.
+     * @return std::nullopt on failure with the relevant Python exception and error set.
+     */
+    [[nodiscard]] static auto create(
+            SchemaTree::Node const* schema_tree_node,
+            std::vector<bool> const& schema_subtree_bitmap,
+            PyDictObject* parent
+    ) -> std::optional<PyDictSerializationIterator> {
+        if (schema_tree_node->is_root() && nullptr != parent) {
+            PyErr_SetString(
+                    PyExc_RuntimeError,
+                    "KeyValuePairLogEvent.to_dict(): Root node cannot have a parent"
+            );
+            return std::nullopt;
+        }
+        if (false == schema_tree_node->is_root() && nullptr == parent) {
+            PyErr_SetString(
+                    PyExc_RuntimeError,
+                    "KeyValuePairLogEvent.to_dict(): Parent cannot be empty for non-root node"
+            );
+            return std::nullopt;
+        }
+
+        PyObjectPtr<PyDictObject> py_dict{py_reinterpret_cast<PyDictObject>(PyDict_New())};
+        if (nullptr == py_dict) {
+            return std::nullopt;
+        }
+
+        std::vector<SchemaTree::Node::id_t> child_schema_tree_nodes;
+        for (auto const child_id : schema_tree_node->get_children_ids()) {
+            if (schema_subtree_bitmap[child_id]) {
+                child_schema_tree_nodes.push_back(child_id);
+            }
+        }
+
+        return PyDictSerializationIterator{
+                schema_tree_node,
+                std::move(child_schema_tree_nodes),
+                parent,
+                std::move(py_dict)
+        };
+    }
+
+    // Delete copy constructor and assignment
+    PyDictSerializationIterator(PyDictSerializationIterator const&) = delete;
+    auto operator=(PyDictSerializationIterator const&) -> PyDictSerializationIterator& = delete;
+
+    // Default move constructor and assignment
+    PyDictSerializationIterator(PyDictSerializationIterator&&) = default;
+    auto operator=(PyDictSerializationIterator&&) -> PyDictSerializationIterator& = default;
+
+    // Destructor
+    ~PyDictSerializationIterator() = default;
+
+    /**
+     * @return Whether there are more child schema tree nodes to traverse.
+     */
+    [[nodiscard]] auto has_next_child_schema_tree_node_id() const -> bool {
+        return m_child_schema_tree_node_it != m_child_schema_tree_nodes.end();
+    }
+
+    /**
+     * Gets the next child schema tree node and advances the iterator.
+     * @return The next child schema tree node.
+     */
+    [[nodiscard]] auto get_next_child_schema_tree_node() -> SchemaTree::Node::id_t {
```

Comment:
You might have added _id to the wrong function, lol

Reply:
yeah, mb. Fixed now